### PR TITLE
carli/2246_scripting_ruler_setSize

### LINKED ISSUE #2246: Scripting: correctly support setSize for a ruler annotation
This is a minor bug / inconsistency in the region class: the `size` property correctly handles the ruler annotation in the same way as a line or vector, by returning the x and y distances between the endpoints. However, the `setSize` function does *not* include the ruler in the line and vector case (which calculates and sets the endpoints using the center point and the new size), and instead falls back to the default case, which sets the second control point to the values provided (which is not the correct interpretation of the control points for this class).

I don't think that this has any effect in the GUI, since this function is not exposed to the user in the edit dialog for this region class, but a scripting user can access this function on the parent class. I have added a workaround to the wrapper (implementing the correct method for this subclass), but a snippet user could run into the same problem, so it would be a good idea to fix this in the frontend.

Edit: `setCenter` has the same issue, and this class also doesn't store a rotation.

### PR DESCRIPTION
**Pull request by crocka**
**Description**

Closes #2246. This PR added ANNRULER to appropriate if-statement in regionStore. Also fixed the failure to update control points when the starting and ending points of a line region have the same x or y value.

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] ~e2e test passing~ / corresponding fix added / ~new e2e test created~
- [x] ~changelog updated~ / no changelog update needed
- [x] ~~protobuf updated to the latest dev commit~~ / no protobuf update needed
- [x] ~~protobuf version bumped~~ / protobuf version not bumped
- [x] `BackendService` unchanged / ~~`BackendService` changed and corresponding ICD test fix added~~

### CODE REVIEW

**Review by YuHsuan-Hwang — APPROVED** [review]
The changes looks good. 👍

**Comment by confluence** [comment]
Are we making a decision for the 2D size of line-shaped regions to have positive components regardless of the direction of the line? This will lead to some unexpected behaviour if the user interacts with region sizes using scripting:

* If the user sets a size with one or more negative components, the size returned with `size` will not be the same as the size set with `setSize`.
* Passing the size returned by `size` back into `setSize` (or another size with positive components) may transform the shape of the region, depending on what the original signs are:
  * If both raw values calculated from the endpoints have been negated, the endpoints will be switched. (Are there any calculations based on line regions where the direction of the endpoints is significant?)
  * If only one of the values has been negated, the region will be rotated by 90°!
* It is no longer possible to specify an orientation for the line by setting the size.

My suggestion would be to keep (half of) the current behaviour, where the 2D size of a line can have negative components, but to flip the endpoints in the calculation -- that is, to subtract endpoint 0 from endpoint 1, so that a component pointing in a positive direction will be positive. Then the orientation of the line will determine the signs of the components, an unambiguous orientation can be selected when the size is set, and the value returned by `size` will preserve the value that was set with `setSize`.

**Comment by confluence** [comment]
Revised idea (as discussed in today's dev meeting): only allow positive sizes, but take the current rotation into account when setting the size, so that the current orientation of the line is preserved.

File: src/stores/Frame/Region/RegionStore.ts [hunk 1]
```diff
@@ -483,8 +482,13 @@ export class RegionStore {
 
     @action setSize = (p: Point2D, skipUpdate = false) => {
         if (this.regionType === CARTA.RegionType.LINE || this.regionType === CARTA.RegionType.ANNLINE || this.regionType === CARTA.RegionType.ANNVECTOR || this.regionType === CARTA.RegionType.ANNRULER) {
-            const newStart = {x: this.center.x - p.x / 2, y: this.center.y - p.y / 2};
-            const newEnd = {x: this.center.x + p.x / 2, y: this.center.y + p.y / 2};
+            const rotation = (this.rotation * Math.PI) / 180.0;
+            const x = Math.abs(p.x);
+            const y = Math.abs(p.y);
+            const _x = x * Math.sign(Math.sin(rotation));
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/Region/RegionStore.ts]: Perhaps rename `_x` to `newX`.

File: src/stores/Frame/Region/RegionStore.ts [hunk 2]
```diff
@@ -483,8 +482,13 @@ export class RegionStore {
 
     @action setSize = (p: Point2D, skipUpdate = false) => {
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/Region/RegionStore.ts:494]: It would be nice if we can add TSDoc to `setSize()` and `size`.

**Review by YuHsuan-Hwang — CHANGES_REQUESTED** [review]
The changes are good.

When the rotation is 0, we fail to set x size. To reproduce with JS snippet:
```js
const file = await app.openFile([path], [any file]);
const region = await file.regionSet.addRegionAsync(1, [{x: 1.5, y: 2}, {x: 1.5, y: 1}]);
//region.setSize({x: 2, y: 3}); // incorrect
//region.setSize({x: 0, y: 3}); // correct
//region.setSize({x: 2, y: 0}); // incorrect
```

File: src/stores/Frame/Region/RegionStore.ts [hunk 3]
```diff
@@ -198,6 +198,10 @@ export class RegionStore {
         }
     }
 
+    /**
+     * Getter for the sizes of regions and annotations
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/Region/RegionStore.ts]: I suggest we revise it to "Returns the sizes of regions and annotations. For line regions and annotations, vector annotations, and ruler annotations, returns x and y absolute displacements."

File: src/stores/Frame/Region/RegionStore.ts [hunk 4]
```diff
@@ -198,6 +198,10 @@ export class RegionStore {
...
+     * @returns A Point2D object with x and y size components.
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/Region/RegionStore.ts]: I suggest we revise it to "The x and y size components." (We don't need to include types.)

File: src/stores/Frame/Region/RegionStore.ts [hunk 5]
```diff
@@ -481,11 +484,23 @@ export class RegionStore {
         }
     };
 
+    /**
+     * Sets the size for regions and annotations
+     *
+     * @param p - Specifies the x and y size components.
+     *            For line region and annotation, vector annotation, and ruler annotation, the function sets the individual new x and y components of the linear segment.
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/Region/RegionStore.ts]: Perhaps revise it to "For line regions and annotations, vector annotations, and ruler annotations, the function sets the new start and end positions while keeping the rotation within the same quadrant."

**Review by YuHsuan-Hwang — CHANGES_REQUESTED** [review]
The new changes looks good to me. I have some comments for the documentation.

Currently, `setSize` updates start/end while keeping the rotation within the same quadrant (with a rotation change <= 90 degrees). @confluence Could you check if this works for scripting?

**Comment by confluence** [comment]
> @confluence Could you check if this works for scripting?

I will check it now, and make the current region PR behave consistently (and adjust the scripting code in dev after the merge / branch).

**Review by confluence — APPROVED** [review]
I'm happy with the current behaviour -- I tested with 4 x 4 regions (one per line-like type x one per quadrant), and confirmed that all the types behave consistently and that orientation is preserved (unless one of the dimensions is set to zero, in which case the orientation is reset to consistent defaults).

I will update the patched functionality in the region PR to match this.

**Review by YuHsuan-Hwang — APPROVED** [review]

**Comment by YuHsuan-Hwang** [comment]
Requires final checks for e2e tests.

**Review by kswang1029 — APPROVED** [review]
tested. a few ruler-related e2e tests have been fixed in a branch.

**Pull request merged**